### PR TITLE
Fast 3DY reduction when the middle dim Y is large but X and Z are small

### PR DESCRIPTION
**Pull request by kaixih**
The existing TF uses a simple CUDA kernel to solve the 3DY reduction. It performs bad when the middle dim Y is large but X and Z are small, because the total thread number is determined by X * Z.

This PR solves this problem by using a new 3DY reduction kernel of two stages: (1) inter-block reduction: starting many blocks to cooperatively reduce the whole dataset by half. This iterates until one block can cover a single or multiple planes; (2) intra-block reduction: each block works independently to reduce the assigned plane(s). For the (2), the shared memory are used for better performance. (Note, for the large X and Z, we will fall back to the original kernel.)

For our user case of N = 16 HW = 1024*1024 C =4 and reduction on HW: before 42.67ms, after 0.97ms, speedup: 44x.

FYI. @nluehr

### CODE REVIEW

**Comment by aaroey** [comment]
@chsigg could you help to take a look at this? Thanks.

**Comment by houtoms** [comment]
Solved a conflict issue.

**Comment by tfboyd** [comment]
Lambda to find the original code owner.

File: tensorflow/core/kernels/reduction_gpu_kernels.cu.h [hunk 1]
```diff
@@ -728,6 +877,72 @@ void Launch3DYReduction(OpKernelContext* ctx, OUT_T out, IN_T in, int extent_x,
...
+
+    n_group_in = n_group_out;
+    temp_in_offset = temp_out_offset;
+    temp_out_offset = temp_in_offset + extent_x * n_group_out * n_size;
+  }
+
+  if (n_group_in * n_size <= threads_per_block) {
+    num_blocks = extent_x;
+  } else { // n_group_in will be 1 here
+    num_blocks =
```
Comment (chsigg) [inline tensorflow/core/kernels/reduction_gpu_kernels.cu.h]: Would this formula also work for the other code path?

**Review by chsigg — COMMENTED** [review]
Sorry for dropping the ball here. I have some mostly stylistic suggestions.

Let me know if you moved on already and would like me to apply them and merge your PR.

**Comment by houtoms** [comment]
@chsigg Thanks for the comments. And I made a set of changes accordingly. Please check. Thx.

**Review by chsigg — APPROVED** [review]

**Comment by aaroey** [comment]
@chsigg thanks for reviewing!

**Pull request merged**